src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1009,6 +1009,15 @@ object Parsers {
               in.nextToken()
               expr()
             } else EmptyTree
+
+          handler match {
+            case Block(Nil, EmptyTree) => syntaxError(
+              "`catch` block does not contain a valid expression, try adding a case like - `case t: Throwable =>` to the block",
```

Comment:
It doesn't:
```scala
scala> try { throw new StackOverflowError } catch { case scala.util.control.NonFatal(e) => println("I caught you!") }
java.lang.StackOverflowError
  at .liftedTree1$1(<console>:8)
  ... 33 elided
```
`NonFatal` is *precisely* meant as a good default, to catch all the things you should catch, and none of those you shouldn't.